### ISSUE TITLE
Should detect whether another WM is already running

### ISSUE THREAD

**Issue by psychon** [sminez/penrose#116]
I just `cargo run --example=minimal` and nothing happened. `xtrace cargo run --example=minimal` shows lots and lots of `BadAccess` errors in response to `GrabKey` and `GrabButton` requests... okay...

However, this piece seems quite relevant:
```
001:<:0080: 16: Request(2): ChangeWindowAttributes window=0x00000761 value-list={event-mask=ButtonMotion,SubstructureNotify,SubstructureRedirect,PropertyChange}
001:>:0080:Error 10=Access: major=2, minor=0, bad=0x00000761, seq=0080
```
I already have a WM running and thus selecting `SubstructureRedirect` failed with an error. I think this is something that you want to catch and error out nicely (libxcb has checked requests for this and I bet xcb-rs also makes them available).

Also, instead of silently ignoring X11 errors in `wait_for_event()`, it might make sense to print them. (Sorry for mixing two issues into one. Feel free to tell me to move this later one to its own report (or to do so yourself).)

**Comment by sminez** [comment]
Yeah, error handling around errors from the X server itself isn't great at the moment / is non-existant. It is probably worthwhile extendin the `XConn` trait to require an explicit method for "check if another WM is running". [dwm has this](https://git.suckless.org/dwm/file/dwm.c.html#l460) and it's just trying to select `SubstructureRedirect` and erroring if that fails. I'll get that put in place and called at the start of `WindowManager::init` so you'll need to pull that in when writing your `XConn` impl :+1:

**Comment by sminez** [comment]
Ok, this works for me for checking that we can set the root event mask:

```rust
fn main() {
    let (conn, _) = xcb::Connection::connect(None).unwrap();
    let root = conn.get_setup().roots().next().unwrap().root();

    let root_event_mask = xcb::EVENT_MASK_PROPERTY_CHANGE
        | xcb::EVENT_MASK_SUBSTRUCTURE_REDIRECT
        | xcb::EVENT_MASK_SUBSTRUCTURE_NOTIFY
        | xcb::EVENT_MASK_BUTTON_MOTION;

    let reply = xcb::change_window_attributes_checked(
        &conn,
        root,
        &[(xcb::CW_EVENT_MASK, root_event_mask)],
    );

    match reply.request_check() {
        Ok(_) => println!("we're good to continue"),
        Err(_) => println!("something else holds substructure notify"),
    }
}
```

I'm going to update the use of these functions to return errors (as they always should have been doing) and get the new method added to `XConn` which will need to be added in your PR #117

**Comment by sminez** [comment]
This is making me realise just how undocumented most of `XConn` is... I'll raise an issue to improve that

**Comment by psychon** [comment]
I did fine just reading the code in `penrose::xcb` and doing the same thing with x11rb. As long as I didn't have to look into `penrose::xcb::conversion`, the things were simple. :-)

**Comment by sminez** [comment]
I'll take your word for it :sweat_smile: 

I think it is probably best to address the X errors now while you are working on your PR: this would mean updating pretty much all of the `XConn` methods to return Results. I'll pick that up as soon as I can (it hopefully shouldn't be _too_ much work :crossed_fingers: )

**Comment by psychon** [comment]
I am fine with `unwrap()`ing for now. What I meant above with "silently ignoring x11 errors" is something like this (dunno if xcb-rs provides any constant to make that hardcoded 0 less ugly, but response type 0 should be this kind of struct: https://docs.rs/xcb/0.9.0/xcb/base/struct.Error.html):
```diff
diff --git a/src/xcb/api.rs b/src/xcb/api.rs
index 3f95923..e22f589 100644
--- a/src/xcb/api.rs
+++ b/src/xcb/api.rs
@@ -358,6 +358,10 @@ impl Api {
                     })
             }
 
+            0 => {
+                warn!("Got an X11 generic error, but right now this code cannot print any details");
+            }
+
             // NOTE: ignoring other event types
             _ => None,
         })
```

**Comment by sminez** [comment]
Implemented in #121

This adds the new unit method for XConn that is now called inside of `WindowManager::init`
